libp2p/transports/tcptransport.nim

```diff
@@ -212,8 +214,30 @@ method dial*(
 
   trace "Dialing remote peer", address = $address
 
-  let transp = await connect(address)
-  return await self.connHandler(transp, Direction.Out)
+  let resolvedAddrs =
```

Comment:
I don't think this is where we want resolution to happen. I can see several reasons as  to why, i) it makes resolution and dialing tightly coupled, ii) it constrains it to one specific transport and if others want to use it, they have to duplicate this logic, ii) it doesn't interact well with a generic dialing engine that will in the future priorities per transport, parallel dialing and potentially other cool things...

I don't think there is a good place for it right now either TBH, so perhaps it makes sense to create an IP based type that all other related transports can derive from - TCP, WS, Quic...

Reply:
Yes, not ideal, but this is required for WSS.

Since we need to resolve the IP in libp2p to be async, and we need to keep the original hostname to send it has a `Host` header & check the certificates, the WS transport needs to have control over the resolution.

Though, I could move the resolution to `connect` in `wire`, and just make the `WSS` transport resolve before calling `connect`.
But it still won't be in the dialer. To move it to the dialer, we should create a new object which holds the hostname + resolved ip. It's a bit cumbersome though

Reply:
> Since we need to resolve the IP in libp2p to be async, and we need to keep the original hostname to send it has a Host header & check the certificates, the WS transport needs to have control over the resolution.

Hmm, you have a point, we need a way to override the hostname in WS tho, we don't have one right now.